File: csharp/Mitch.cs
using System.Buffers.Binary;
using System.Runtime.InteropServices;

namespace NxRates.Sdk;

// ---------------------------------------------------------------------------
//  Constants
// ---------------------------------------------------------------------------

public static class MitchConstants
{
    public const int HeaderSize = 16;
    public const int TickSize = 32;
    public const int TradeSize = 24;
    public const int IndexSize = 40;

    // ASCII message-type bytes
    public const byte MsgTrade     = (byte)'t';
    public const byte MsgOrder     = (byte)'o';
    public const byte MsgTick      = (byte)'s';
    public const byte MsgIndex     = (byte)'i';
    public const byte MsgOrderBook = (byte)'b';
    public const byte MsgBar       = (byte)'k';

    // Wire code <-> ASCII lookup
    private static readonly byte[] WireToAscii = { 0, MsgTrade, MsgOrder, MsgTick, MsgIndex, MsgOrderBook, MsgBar };
    private static readonly byte[] AsciiToWire = new byte[256];

    static MitchConstants()
    {
        for (int i = 1; i < WireToAscii.Length; i++)
            AsciiToWire[WireToAscii[i]] = (byte)i;
    }

    public static byte WireCodeToAscii(int wireCode) =>
        wireCode >= 1 && wireCode < WireToAscii.Length ? WireToAscii[wireCode] : (byte)0;

    public static byte AsciiToWireCode(byte ascii) => AsciiToWire[ascii];
}

// ---------------------------------------------------------------------------
//  Timestamp helpers
// ---------------------------------------------------------------------------

public static class MitchTimestamp
{
    /// <summary>Microseconds from Unix epoch to 2010-01-01T00:00:00Z.</summary>
    public const long Epoch2010Us = 1_262_304_000_000_000L;

    private const long TicksPerMicrosecond = 16;

    /// <summary>Decode a 48-bit LE tick count to a DateTimeOffset (UTC).</summary>
    public static DateTimeOffset FromEpochUs(long ticks16us)
    {
        // Each tick = 16 us. Total microseconds since 2010 = ticks * 16.
        long us = ticks16us * TicksPerMicrosecond;
        long unixUs = us + Epoch2010Us;
        long dotnetTicks = unixUs * 10; // 1 us = 10 .NET ticks (100ns each)
        return new DateTimeOffset(dotnetTicks + DateTimeOffset.UnixEpoch.Ticks, TimeSpan.Zero);
    }

    /// <summary>Encode a DateTimeOffset to a 48-bit tick count (16 us resolution).</summary>
    public static long ToEpochUs(DateTimeOffset dt)
    {
        long dotnetTicks = dt.UtcTicks - DateTimeOffset.UnixEpoch.Ticks;
        long unixUs = dotnetTicks / 10;
        long usSince2010 = unixUs - Epoch2010Us;
        return usSince2010 / TicksPerMicrosecond;
    }

    internal static long ReadU48LE(ReadOnlySpan<byte> src)
    {
        // Read 6 bytes little-endian into a long.
        long v = src[0]
               | ((long)src[1] << 8)
               | ((long)src[2] << 16)
               | ((long)src[3] << 24)
               | ((long)src[4] << 32)
               | ((long)src[5] << 40);
        return v;
    }

    internal static void WriteU48LE(Span<byte> dst, long value)
    {
        dst[0] = (byte)(value);
        dst[1] = (byte)(value >> 8);
        dst[2] = (byte)(value >> 16);
        dst[3] = (byte)(value >> 24);
        dst[4] = (byte)(value >> 32);
        dst[5] = (byte)(value >> 40);
    }
}

// ---------------------------------------------------------------------------
//  MitchHeader (16 bytes)
// ---------------------------------------------------------------------------

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MitchHeader
{
    public const int Size = MitchConstants.HeaderSize;

    public ushort TypeProvider;  // [3:0]=wire code, [15:4]=provider_id
    public long   TimestampRaw; // only low 48 bits used on wire
    public byte   Count;
    public byte   Flags;
    public ushort Sequence;
    public uint   Reserved;

    // Decoded properties -------------------------------------------------------

    /// <summary>ASCII message type ('t','o','s','i','b','k').</summary>
    public readonly byte MsgType => MitchConstants.WireCodeToAscii(TypeProvider & 0x0F);

    /// <summary>Provider id (0-4095).</summary>
    public readonly ushort ProviderId => (ushort)(TypeProvider >> 4);

    /// <summary>Decoded UTC timestamp.</summary>
    public readonly DateTimeOffset Timestamp => MitchTimestamp.FromEpochUs(TimestampRaw);

    // Wire I/O -----------------------------------------------------------------

    public readonly void WriteTo(Span<byte> dst)
    {
        BinaryPrimitives.WriteUInt16LittleEndian(dst, TypeProvider);
        MitchTimestamp.WriteU48LE(dst.Slice(2), TimestampRaw);
        dst[8] = Count;
        dst[9] = Flags;
        BinaryPrimitives.WriteUInt16LittleEndian(dst.Slice(10), Sequence);
        BinaryPrimitives.WriteUInt32LittleEndian(dst.Slice(12), Reserved);
    }

    public static MitchHeader ReadFrom(ReadOnlySpan<byte> src)
    {
        return new MitchHeader
        {
            TypeProvider = BinaryPrimitives.ReadUInt16LittleEndian(src),
            TimestampRaw = MitchTimestamp.ReadU48LE(src.Slice(2)),
            Count        = src[8],
            Flags        = src[9],
            Sequence     = BinaryPrimitives.ReadUInt16LittleEndian(src.Slice(10)),
            Reserved     = BinaryPrimitives.ReadUInt32LittleEndian(src.Slice(12)),
        };
    }

    // Factory helpers -----------------------------------------------------------

    public static ushort PackTypeProvider(byte asciiType, ushort providerId) =>
        (ushort)((providerId << 4) | MitchConstants.AsciiToWireCode(asciiType));
}

// ---------------------------------------------------------------------------
//  Tick (32 bytes)
// ---------------------------------------------------------------------------

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MitchTick
{
    public const int Size = MitchConstants.TickSize;

    public ulong  Ticker;  // @0
    public double Bid;     // @8
    public double Ask;     // @16
    public uint   VBid;    // @24
    public uint   VAsk;    // @28

    public readonly void WriteTo(Span<byte> dst)
    {
        BinaryPrimitives.WriteUInt64LittleEndian(dst, Ticker);
        BinaryPrimitives.WriteDoubleLittleEndian(dst.Slice(8), Bid);
        BinaryPrimitives.WriteDoubleLittleEndian(dst.Slice(16), Ask);
        BinaryPrimitives.WriteUInt32LittleEndian(dst.Slice(24), VBid);
        BinaryPrimitives.WriteUInt32LittleEndian(dst.Slice(28), VAsk);
    }

    public static MitchTick ReadFrom(ReadOnlySpan<byte> src) => new()
    {
        Ticker = BinaryPrimitives.ReadUInt64LittleEndian(src),
        Bid    = BinaryPrimitives.ReadDoubleLittleEndian(src.Slice(8)),
        Ask    = BinaryPrimitives.ReadDoubleLittleEndian(src.Slice(16)),
        VBid   = BinaryPrimitives.ReadUInt32LittleEndian(src.Slice(24)),
        VAsk   = BinaryPrimitives.ReadUInt32LittleEndian(src.Slice(28)),
    };
}

// ---------------------------------------------------------------------------
//  Trade (24 bytes)
// ---------------------------------------------------------------------------

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MitchTrade
{
    public const int Size = MitchConstants.TradeSize;

    public ulong  Ticker;   // @0
    public double Price;    // @8
    public uint   Qty;      // @16
    // TradeId is u24 LE @20 (3 bytes), Side is u8 @23
    public uint   TradeIdAndSide; // packing helper -- low 24 bits = TradeId, byte 3 = Side

    public uint TradeId
    {
        readonly get => TradeIdAndSide & 0x00FF_FFFF;
        set => TradeIdAndSide = (TradeIdAndSide & 0xFF00_0000) | (value & 0x00FF_FFFF);
    }

    public byte Side
    {
        readonly get => (byte)(TradeIdAndSide >> 24);
        set => TradeIdAndSide = (TradeIdAndSide & 0x00FF_FFFF) | ((uint)value << 24);
    }

    public readonly void WriteTo(Span<byte> dst)
    {
        BinaryPrimitives.WriteUInt64LittleEndian(dst, Ticker);
        BinaryPrimitives.WriteDoubleLittleEndian(dst.Slice(8), Price);
        BinaryPrimitives.WriteUInt32LittleEndian(dst.Slice(16), Qty);
        // TradeId: 3 bytes LE @20
        uint tid = TradeId;
        dst[20] = (byte)(tid);
        dst[21] = (byte)(tid >> 8);
        dst[22] = (byte)(tid >> 16);
        dst[23] = Side;
    }

    public static MitchTrade ReadFrom(ReadOnlySpan<byte> src)
    {
        uint tid = (uint)(src[20] | (src[21] << 8) | (src[22] << 16));
        byte side = src[23];
        return new MitchTrade
        {
            Ticker         = BinaryPrimitives.ReadUInt64LittleEndian(src),
            Price          = BinaryPrimitives.ReadDoubleLittleEndian(src.Slice(8)),
            Qty            = BinaryPrimitives.ReadUInt32LittleEndian(src.Slice(16)),
            TradeIdAndSide = tid | ((uint)side << 24),
        };
    }
}

// ---------------------------------------------------------------------------
//  Index (40 bytes)
// ---------------------------------------------------------------------------

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MitchIndex
{
    public const int Size = MitchConstants.IndexSize;

    public ulong  Ticker;     // @0
    public double Bid;        // @8
    public double Ask;        // @16
    public uint   VBid;       // @24
    public uint   VAsk;       // @28
    public ushort CI;         // @32
    public ushort TickCount;  // @34
    public byte   Confidence; // @36
    public byte   Accepted;   // @37
    public byte   Rejected;   // @38
    public byte   Pad;        // @39

    public readonly void WriteTo(Span<byte> dst)
    {
        BinaryPrimitives.WriteUInt64LittleEndian(dst, Ticker);
        BinaryPrimitives.WriteDoubleLittleEndian(dst.Slice(8), Bid);
        BinaryPrimitives.WriteDoubleLittleEndian(dst.Slice(16), Ask);
        BinaryPrimitives.WriteUInt32LittleEndian(dst.Slice(24), VBid);
        BinaryPrimitives.WriteUInt32LittleEndian(dst.Slice(28), VAsk);
        BinaryPrimitives.WriteUInt16LittleEndian(dst.Slice(32), CI);
        BinaryPrimitives.WriteUInt16LittleEndian(dst.Slice(34), TickCount);
        dst[36] = Confidence;
        dst[37] = Accepted;
        dst[38] = Rejected;
        dst[39] = Pad;
    }

    public static MitchIndex ReadFrom(ReadOnlySpan<byte> src) => new()
    {
        Ticker     = BinaryPrimitives.ReadUInt64LittleEndian(src),
        Bid        = BinaryPrimitives.ReadDoubleLittleEndian(src.Slice(8)),
        Ask        = BinaryPrimitives.ReadDoubleLittleEndian(src.Slice(16)),
        VBid       = BinaryPrimitives.ReadUInt32LittleEndian(src.Slice(24)),
        VAsk       = BinaryPrimitives.ReadUInt32LittleEndian(src.Slice(28)),
        CI         = BinaryPrimitives.ReadUInt16LittleEndian(src.Slice(32)),
        TickCount  = BinaryPrimitives.ReadUInt16LittleEndian(src.Slice(34)),
        Confidence = src[36],
        Accepted   = src[37],
        Rejected   = src[38],
        Pad        = src[39],
    };
}

File: java/src/main/java/com/nxrates/sdk/Mitch.java
package com.nxrates.sdk;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * MITCH wire format types and codec.
 * All multi-byte fields are little-endian. Java records for zero-overhead value types.
 */
public final class Mitch {

    private Mitch() {}

    // ── sizes ────────────────────────────────────────────────────────────
    public static final int SIZE_HEADER     = 16;
    public static final int SIZE_TRADE      = 24;
    public static final int SIZE_ORDER      = 32;
    public static final int SIZE_TICK       = 32;
    public static final int SIZE_INDEX      = 40;
    public static final int SIZE_BAR        = 128;
    public static final int SIZE_ORDER_BOOK = 2072;

    // ── message type ASCII codes ─────────────────────────────────────────
    public static final byte MSG_TRADE      = 't';
    public static final byte MSG_ORDER      = 'o';
    public static final byte MSG_TICK       = 's';
    public static final byte MSG_INDEX      = 'i';
    public static final byte MSG_BAR        = 'k';
    public static final byte MSG_ORDER_BOOK = 'b';

    // ── wire code <-> ASCII mapping ──────────────────────────────────────
    private static final byte[] WIRE_TO_ASCII = {0, MSG_TRADE, MSG_ORDER, MSG_TICK, MSG_INDEX, MSG_ORDER_BOOK, MSG_BAR};
    //                                           0      1          2          3         4            5           6

    /** Convert wire code (1-6) to ASCII message type. */
    public static byte wireCodeToAscii(int wireCode) {
        if (wireCode < 1 || wireCode > 6) throw new IllegalArgumentException("wire code out of range: " + wireCode);
        return WIRE_TO_ASCII[wireCode];
    }

    /** Convert ASCII message type to wire code (1-6). */
    public static int asciiToWireCode(byte ascii) {
        return switch (ascii) {
            case MSG_TRADE      -> 1;
            case MSG_ORDER      -> 2;
            case MSG_TICK       -> 3;
            case MSG_INDEX      -> 4;
            case MSG_ORDER_BOOK -> 5;
            case MSG_BAR        -> 6;
            default -> throw new IllegalArgumentException("unknown msg type: " + (char) ascii);
        };
    }

    // ── timestamp ────────────────────────────────────────────────────────
    /** Microseconds from Unix epoch to 2010-01-01T00:00:00Z. */
    public static final long EPOCH_2010_US = 1_262_304_000_000_000L;

    /** Encode a Unix-epoch microsecond timestamp to a u48 MITCH tick value (16 us granularity). */
    public static long fromEpochUs(long epochUs) {
        return (epochUs - EPOCH_2010_US) / 16;
    }

    /** Decode a u48 MITCH tick value back to Unix-epoch microseconds. */
    public static long toEpochUs(long ticks) {
        return ticks * 16 + EPOCH_2010_US;
    }

    /** Write a u48 LE value into buf at the current position (advances position by 6). */
    static void putU48(ByteBuffer buf, long value) {
        // write low 4 bytes then high 2 bytes
        buf.putInt((int) (value & 0xFFFFFFFFL));
        buf.putShort((short) ((value >>> 32) & 0xFFFFL));
    }

    /** Read a u48 LE value from buf at the current position (advances position by 6). */
    static long getU48(ByteBuffer buf) {
        long low  = Integer.toUnsignedLong(buf.getInt());
        long high = Short.toUnsignedLong(buf.getShort());
        return low | (high << 32);
    }

    // ── MitchHeader ──────────────────────────────────────────────────────
    /**
     * MITCH v2 header (16 bytes).
     *
     * @param typeProvider packed u16: bits [3:0] = wire code, bits [15:4] = provider_id
     * @param timestamp    u48 LE, 16us ticks since 2010-01-01
     * @param count        batch entry count (1-255)
     * @param flags        [1:0] = version, [7:2] = reserved
     * @param sequence     per-stream gap detection
     * @param reserved     4 reserved bytes
     */
    public record MitchHeader(
            short typeProvider,
            long  timestamp,
            int   count,
            int   flags,
            int   sequence,
            byte[] reserved
    ) {
        /** Decode the wire code to an ASCII message type char. */
        public byte msgType() {
            int wireCode = Short.toUnsignedInt(typeProvider) & 0x0F;
            return wireCodeToAscii(wireCode);
        }

        /** Extract the provider id (bits 4-15 of typeProvider). */
        public int providerId() {
            return (Short.toUnsignedInt(typeProvider) >>> 4) & 0x0FFF;
        }

        /** Build typeProvider from ASCII msg type and provider id. */
        public static short packTypeProvider(byte asciiType, int providerId) {
            int wireCode = asciiToWireCode(asciiType);
            return (short) ((wireCode & 0x0F) | ((providerId & 0x0FFF) << 4));
        }

        public byte[] pack() {
            var buf = ByteBuffer.allocate(SIZE_HEADER).order(ByteOrder.LITTLE_ENDIAN);
            buf.putShort(typeProvider);
            putU48(buf, timestamp);
            buf.put((byte) (count & 0xFF));
            buf.put((byte) (flags & 0xFF));
            buf.putShort((short) (sequence & 0xFFFF));
            byte[] res = reserved != null ? reserved : new byte[4];
            buf.put(res, 0, 4);
            return buf.array();
        }

        public static MitchHeader unpack(byte[] data) {
            return unpack(ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN));
        }

        public static MitchHeader unpack(ByteBuffer buf) {
            buf.order(ByteOrder.LITTLE_ENDIAN);
            short tp    = buf.getShort();
            long  ts    = getU48(buf);
            int   cnt   = Byte.toUnsignedInt(buf.get());
            int   fl    = Byte.toUnsignedInt(buf.get());
            int   seq   = Short.toUnsignedInt(buf.getShort());
            byte[] res  = new byte[4];
            buf.get(res);
            return new MitchHeader(tp, ts, cnt, fl, seq, res);
        }
    }

    // ── Tick (32 bytes) ──────────────────────────────────────────────────
    /**
     * @param ticker encoded instrument id
     * @param bid    best bid price
     * @param ask    best ask price
     * @param vbid   bid volume
     * @param vask   ask volume
     */
    public record Tick(long ticker, double bid, double ask, int vbid, int vask) {

        public byte[] pack() {
            var buf = ByteBuffer.allocate(SIZE_TICK).order(ByteOrder.LITTLE_ENDIAN);
            buf.putLong(ticker);
            buf.putDouble(bid);
            buf.putDouble(ask);
            buf.putInt(vbid);
            buf.putInt(vask);
            return buf.array();
        }

        public static Tick unpack(byte[] data) {
            return unpack(ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN));
        }

        public static Tick unpack(ByteBuffer buf) {
            buf.order(ByteOrder.LITTLE_ENDIAN);
            return new Tick(buf.getLong(), buf.getDouble(), buf.getDouble(), buf.getInt(), buf.getInt());
        }
    }

    // ── Trade (24 bytes) ─────────────────────────────────────────────────
    /**
     * @param ticker  encoded instrument id
     * @param price   trade price
     * @param qty     trade quantity
     * @param tradeId 24-bit trade id (0 .. 16_777_215)
     * @param side    trade side byte
     */
    public record Trade(long ticker, double price, int qty, int tradeId, byte side) {

        public byte[] pack() {
            var buf = ByteBuffer.allocate(SIZE_TRADE).order(ByteOrder.LITTLE_ENDIAN);
            buf.putLong(ticker);
            buf.putDouble(price);
            buf.putInt(qty);
            // tradeId: 3 bytes LE at offset 20
            buf.put((byte) (tradeId & 0xFF));
            buf.put((byte) ((tradeId >>> 8) & 0xFF));
            buf.put((byte) ((tradeId >>> 16) & 0xFF));
            buf.put(side);
            return buf.array();
        }

        public static Trade unpack(byte[] data) {
            return unpack(ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN));
        }

        public static Trade unpack(ByteBuffer buf) {
            buf.order(ByteOrder.LITTLE_ENDIAN);
            long   tk  = buf.getLong();
            double px  = buf.getDouble();
            int    q   = buf.getInt();
            int    id  = Byte.toUnsignedInt(buf.get())
                       | (Byte.toUnsignedInt(buf.get()) << 8)
                       | (Byte.toUnsignedInt(buf.get()) << 16);
            byte   s   = buf.get();
            return new Trade(tk, px, q, id, s);
        }
    }

    // ── Index (40 bytes) ─────────────────────────────────────────────────
    /**
     * @param ticker     encoded instrument id
     * @param bid        index bid
     * @param ask        index ask
     * @param vbid       bid volume
     * @param vask       ask volume
     * @param ci         composite index value (short)
     * @param tickCount  number of ticks (short)
     * @param confidence confidence level (byte)
     * @param accepted   accepted count (byte)
     * @param rejected   rejected count (byte)
     * @param pad        padding byte
     */
    public record Index(
            long ticker, double bid, double ask,
            int vbid, int vask,
            short ci, short tickCount,
            byte confidence, byte accepted, byte rejected, byte pad
    ) {
        public byte[] pack() {
            var buf = ByteBuffer.allocate(SIZE_INDEX).order(ByteOrder.LITTLE_ENDIAN);
            buf.putLong(ticker);
            buf.putDouble(bid);
            buf.putDouble(ask);
            buf.putInt(vbid);
            buf.putInt(vask);
            buf.putShort(ci);
            buf.putShort(tickCount);
            buf.put(confidence);
            buf.put(accepted);
            buf.put(rejected);
            buf.put(pad);
            return buf.array();
        }

        public static Index unpack(byte[] data) {
            return unpack(ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN));
        }

        public static Index unpack(ByteBuffer buf) {
            buf.order(ByteOrder.LITTLE_ENDIAN);
            long   tk   = buf.getLong();
            double b    = buf.getDouble();
            double a    = buf.getDouble();
            int    vb   = buf.getInt();
            int    va   = buf.getInt();
            short  ci   = buf.getShort();
            short  tc   = buf.getShort();
            byte   conf = buf.get();
            byte   acc  = buf.get();
            byte   rej  = buf.get();
            byte   pad  = buf.get();
            return new Index(tk, b, a, vb, va, ci, tc, conf, acc, rej, pad);
        }
    }
}
